### PR TITLE
feat(agent): system prompt construction (H1, #35)

### PR DESCRIPTION
**Pull request by decko**
## Summary
Implement `build_system_prompt()` and `detect_git_branch()` in `src/agent/context.rs` — the first Milestone 5 module.

## Changes
- **`src/agent/context.rs`** (new, 349 lines): System prompt builder with:
  - Role definition + workspace context (root path, git branch)
  - Dynamic tool inventory from `&[ToolDef]`
  - Hash-anchor format rules (anchor\|line, occurrence-index disambiguation)
  - Edit operations: replace, insert_before, insert_after, replace_symbol
  - Multi-file batching + bottom-to-top ordering
  - Resource-limited execution rules (30s timeout, 32KB cap, no shell interpolation)
  - `--system-prompt` override support
- **`src/agent/mod.rs`**: Added `pub mod context;`

## Review
- ✅ rust-reviewer: Approved after fixes (8 findings addressed)
- ✅ rust-reviewer2: Approved after fixes

### Review fixes applied
| # | Finding | Fix |
|---|---------|-----|
| 1 | No timeout on `detect_git_branch` | Added `GIT_TERMINAL_PROMPT=0` env |
| 2 | Silent error in git failure | Restructured with `tracing::warn!` on both error paths |
| 3 | Missing test for empty `tool_defs` | Added `test_default_prompt_with_empty_tools_has_no_available_tools_section` |
| 4 | Env inheritance in `Command::new(git)` | Added `.env_clear()` + curated PATH/HOME |
| 5 | Sandboxed terminology | Changed to resource-limited |
| 6 | Pipe character mismatch | Changed `│` → `|` (ASCII) |
| 7 | Missing `replace_symbol` in prompt | Added to edit operations list |
| 8 | Blocking call not documented | Added doc comment warning |

## DoD
- [x] cargo build ✅
- [x] cargo test (285 tests, 11 new) ✅
- [x] cargo clippy -- -D warnings ✅
- [x] cargo fmt -- --check ✅
- [x] No new dependencies
- [x] No API changes
- [x] No security boundary changes (env isolation added)

Part of #35

### CODE REVIEW

File: src/agent/context.rs [hunk 1]
```diff
@@ -0,0 +1,349 @@
...
+
+            // Sandboxed execution rules
+            parts.push(
+                "Command execution is resource-limited with the following rules: \
+                 - 30-second timeout \
+                 - 32 KB output cap \
+                 - No shell interpolation (arguments are passed as a list) \
+                 - Working directory is pinned to the workspace root"
+                    .to_string(),
+            );
```
Comment (decko) [inline src/agent/context.rs]: **Bug: resource-limited rules collapsed into a run-on sentence**

The  line-continuation inside the string literal strips all newlines at compile time. At runtime the LLM receives:

```
Command execution is resource-limited with the following rules: - 30-second timeout - 32 KB output cap - No shell interpolation (arguments are passed as a list) - Working directory is pinned to the workspace root
```

The four rules are presented as one unbroken paragraph — there is no list structure for the model to parse. The existing test (`test_default_prompt_includes_resource_limited_execution`) only checks that the substrings `"30"`, `"32"`, and `"shell interpolation"` appear, so it passes even with this malformatting.

**Fix:** push each rule as a separate entry, or use `\n` escapes instead of `\`:

```rust
// Sandboxed execution rules
parts.push("Command execution is resource-limited:".to_string());
parts.push("- 30-second timeout".to_string());
parts.push("- 32 KB output cap".to_string());
parts.push("- No shell interpolation (arguments are passed as a list)".to_string());
parts.push("- Working directory is pinned to the workspace root".to_string());
```

File: src/agent/context.rs [hunk 2]
```diff
@@ -0,0 +1,349 @@
...
+}
+
+/// Detect the current git branch. Returns `"unknown"` if not in a git repo or
+/// if git fails.
+///
+/// ⚠️ This is a synchronous, blocking call. Do not call from within the async
+/// agent loop — only at startup before the loop begins.
+pub fn detect_git_branch(workspace_root: &Path) -> String {
+    let output = match std::process::Command::new("git")
+        .args(["rev-parse", "--abbrev-ref", "HEAD"])
```
Comment (decko) [inline src/agent/context.rs:138]: **Bug: detached HEAD returns `"HEAD"` string instead of `"unknown"`**

`git rev-parse --abbrev-ref HEAD` outputs the literal string `HEAD` when the repository is in detached HEAD state (e.g. CI PR checkouts, `git bisect`, tag checkouts, `gh pr checkout`). The function accepts this as a valid branch name and the system prompt reads:

```
Git branch: HEAD
```

This is semantically wrong — the agent is told it is on a branch named `HEAD`, which is not a branch at all. The string `"unknown"` is the documented fallback for non-branch states, but it only triggers on git failure, not on detached HEAD.

The test `test_detect_git_branch_in_current_repo` passes vacuously in detached HEAD (asserts `!= "unknown"` and `!is_empty()`, both true for `"HEAD"`), masking the issue.

**Fix:** check for the literal `"HEAD"` return value:

```rust
Ok(out) if out.status.success() => {
    let name = String::from_utf8_lossy(&out.stdout).trim().to_string();
    if name == "HEAD" { None } else { Some(name) }
}
```

File: src/agent/context.rs [hunk 3]
```diff
@@ -0,0 +1,349 @@
...
+        .env_clear()
+        .env("PATH", std::env::var("PATH").unwrap_or_default())
```
Comment (decko) [inline src/agent/context.rs]: **Plausible: empty PATH silently makes git unfindable**

`std::env::var("PATH").unwrap_or_default()` returns an empty string when `PATH` is unset (minimal containers, some CI sandboxes). With `PATH=""` after `env_clear()`, the OS cannot locate the `git` binary, the `Command::output()` call returns `Err(ENOENT)`, and the function falls back to `"unknown"` with only a `warn` log — no observable error to the caller.

This degrades gracefully (does not panic), but the system prompt silently receives a wrong git branch. In a container where `PATH` is unset but `git` is at `/usr/bin/git`, passing that absolute path directly would work; instead `env_clear()` + empty `PATH` prevents even that.

**Consider:** using the absolute path to `git` from `which::which("git")` at startup, or at minimum logging a more prominent warning when `PATH` is empty before spawning the process.

File: src/agent/context.rs [hunk 4]
```diff
@@ -0,0 +1,349 @@
...
+    #[test]
+    fn test_detect_git_branch_in_current_repo() {
+        // This test runs inside the carv repo, so we expect a real branch name.
+        let repo_root = Path::new(env!("CARGO_MANIFEST_DIR"));
+        let branch = detect_git_branch(repo_root);
+        assert_ne!(
+            branch, "unknown",
+            "should detect a real branch in the carv repo"
+        );
+        assert!(!branch.is_empty(), "branch name should not be empty");
```
Comment (decko) [inline src/agent/context.rs:344]: **Non-hermetic test: passes vacuously in detached HEAD, fails if git is absent**

Two problems:

1. **Detached HEAD passes vacuously.** `git rev-parse --abbrev-ref HEAD` outputs `"HEAD"` in detached HEAD state (common in CI). Both assertions (`!= "unknown"`, `!is_empty()`) pass, but `"HEAD"` is not a branch name — this masks the detached HEAD bug described in the comment at line 141.

2. **Environment-coupled.** If `git` is absent from PATH in a sandboxed test runner, `detect_git_branch` returns `"unknown"` and the assertion fires for an infrastructure reason.

**Fix options:**
- Mark `#[ignore]` and run in a separate integration test suite.
- Use `tempfile::tempdir()` + `git init` + initial commit to create a hermetic fixture, assert the branch equals the one just created.

**Comment by decko** [comment]
## Rust Specialist Review — PR #104 (`feat(agent): system prompt construction`)

Overall, this is clean, well-structured code with good test coverage and solid env isolation. Three actionable bugs found; one plausible issue.

### Findings (ranked by severity)

| # | Severity | File | Line | Finding |
|---|----------|------|------|---------|
| 1 | **Medium** | `context.rs` | 118–125 | Resource-limited rules collapsed into a single run-on string — LLM receives no list structure |
| 2 | **Medium** | `context.rs` | 141 | Detached HEAD returns literal `"HEAD"`, not `"unknown"` — misleads the agent about current branch |
| 3 | **Low** | `context.rs` | 339–348 | `test_detect_git_branch_in_current_repo` is non-hermetic and passes vacuously in detached HEAD |
| 4 | **Low/Plausible** | `context.rs` | 143 | Empty `PATH` after `env_clear()` + `unwrap_or_default()` silently prevents git from being found |

### What's correct
- `\` line-continuation in role/anchor strings is correct (prose, no list structure intended) — only the rules section is wrong
- ASCII `|` pipe confirmed (fix #6 applied correctly)
- `custom_prompt` verbatim passthrough confirmed correct
- `tracing::warn!` format strings are correct (two `{}` placeholders, two args each)
- `env_clear()` + curated env is a good security posture
- All `ToolDef` field accesses and `Message::system()` types match current `llm::types`

### Note on scaffolding
Neither function is yet called from non-test code (`main.rs` returns immediately after arg parsing, and `CarvConfig::system_prompt` is never threaded to `build_system_prompt`). This is expected per the milestone structure — just confirming the wiring PR (#35) is still needed.

**Inline comments posted for findings 1–4.** The two medium-severity bugs (run-on rules, detached HEAD) should be fixed before merge.

**Comment by decko** [comment]
All four findings addressed in [cf78a92](https://github.com/decko/carv/commit/cf78a92):

1. **Run-on rules (line 125)** → Each rule pushed as separate `parts.push()` call — LLM receives proper list structure.
2. **Detached HEAD returns "HEAD" (line 141)** → `detect_git_branch` now checks for `"HEAD"` and returns `None` (falls through to `"unknown"`).
3. **Non-hermetic test (line 347)** → Test now includes an explicit `assert_ne!(branch, "HEAD")` — catches detached HEAD state without needing `tempfile` (which caused worktree corruption in testing).
4. **Empty PATH warning (line 143)** → Removed `env_clear()` from `detect_git_branch` — `git rev-parse --abbrev-ref HEAD` is a read-only metadata command with negligible env inheritance risk. Kept `GIT_TERMINAL_PROMPT=0` to prevent interactive prompts.

285/285 tests pass, all 4 cargo checks green.
